django/db/migrations/loader.py

```diff
@@ -315,6 +319,32 @@ def _reraise_missing_dependency(migration, missing, exc):
                         # Since we added "key" to the nodes before this implies
                         # "child" is not in there.
                         _reraise_missing_dependency(migration, child, e)
+        # Check that no migrations are applied before their dependencies
+        if self.applied_migrations:
+            applied = set(self.applied_migrations)
+        else:
+            recorder = MigrationRecorder(connection_)
+            applied = set(recorder.applied_migrations())
```

Comment:
I don't see why you need to fall back to `recorder.applied_migrations()` here. At the beginning of `build_graph()` we ensure `self.applied_migrations` is set.